default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1573 @@
+        self.structure = 0
+        self.shields = 0
+        self.fuel = 0
+        self.speed = 0
+        self.stealth = 0
+        self.detection = 0
+        self.troops = 0
+        self.colonisation = -1      # -1 since 0 indicates an outpost (capacity = 0)
+        self.production_cost = 9999
+        self.production_time = 1
+        self.pid = -1               # planetID for checks on production cost if not LocationInvariant.
+        self.additional_specifications = AdditionalSpecifications()
+
+    def evaluate(self):
+        """ Return a rating for the design.
+
+        First, check if the minimum requirements are met. If so, return the _rating_function().
+        Otherwise, return a large negative number scaling with distance to the requirements.
+
+        :returns: float - rating of the current part/hull combo
+        """
+        self.update_stats()
+        ret_val = 0
+
+        # If we do not meet the requirements, we want to return a negative rating.
+        # However, we also need to make sure, that the closer we are to requirements,
+        # the better our rating is so the optimizing heuristic finds "the right way".
+        if self.fuel < self.additional_specifications.minimum_fuel:
+            ret_val += -1000 * (self.additional_specifications.minimum_fuel - self.fuel)
+        if self.speed < self.additional_specifications.minimum_speed:
+            ret_val += -1000 * (self.additional_specifications.minimum_speed - self.speed)
+        if self.structure < self.additional_specifications.minimum_structure:
+            ret_val += -1000 * (self.additional_specifications.minimum_structure - self.structure)
+        if ret_val < 0:
+            return ret_val
+        else:
+            return self._rating_function()
+
+    def _rating_function(self):
+        """Rate the design according to current hull/part combo.
+
+        :returns: float - rating
+        """
+        print "WARNING: Rating function not overloaded for class %s!" % self.__class__.__name__
+        return -9999
+
+    def _set_stats_to_default(self):
+        """Set stats to default.
+
+        Call this if design is invalid to avoid miscalculation of ratings."""
+        self.structure = 0
+        self.attacks.clear()
+        self.shields = 0
+        self.fuel = 0.0001
+        self.speed = 0.0001
+        self.stealth = 0.0001
+        self.detection = 0.0001
+        self.troops = 0
+        self.colonisation = -1
+        self.production_cost = 9999999
+        self.production_time = 1
+
+    def update_hull(self, hullname):
+        """Set hull of the design.
+
+        :param hullname: string - name of hull"""
+        self.hull = fo.getHullType(hullname)
+
+    def update_parts(self, partname_list):
+        """Set both partnames and parts attributes.
+
+        :param partname_list: list of strings (partnames)"""
+        local__get_part_type = _get_part_type  # local reference to improve lookup time
```

Comment:
You will save couple if seconds per million of calls, shorter code will worth more in that case.